Mathlib/GroupTheory/Coxeter/Basic.lean

```diff
@@ -166,4 +186,179 @@ theorem _root_.CoxeterMatrix.toCoxeterSystem_simple (M : CoxeterMatrix B) :
+    fun w _ ↦ p w
+  have := cs.submonoid_closure_range_simple.symm ▸ Submonoid.mem_top w
+  apply Submonoid.closure_induction_left (p := p')
+  · exact one
+  · rintro _ ⟨i, rfl⟩ y _
+    exact mul_simple_left y i
+  · exact this
+
+/-- If `p : W → Prop` holds for the identity and it is preserved under multiplying on the right
+by a simple reflection, then it holds for all elements of `W`. -/
+theorem simple_induction_right {p : W → Prop} (w : W) (one : p 1)
+    (mul_simple_right : ∀ (w : W) (i : B), p w → p (w * s i)) : p w := by
+  let p' : ((w : W) → w ∈ Submonoid.closure (Set.range cs.simple) → Prop) :=
+    fun w _ ↦ p w
+  have := cs.submonoid_closure_range_simple.symm ▸ Submonoid.mem_top w
+  apply Submonoid.closure_induction_right (p := p')
+  · exact one
+  · rintro x _ _ ⟨i, rfl⟩
+    exact mul_simple_right x i
+  · exact this
+
+/-! ### Homomorphisms from a Coxeter group -/
+
+/-- The proposition that the values of the function `f : B → G` satisfy the Coxeter relations
+corresponding to the matrix `M`. -/
+def _root_.CoxeterMatrix.IsLiftable {G : Type*} [Monoid G] (M : CoxeterMatrix B) (f : B → G) :
+    Prop := ∀ i i', (f i * f i') ^ M i i' = 1
+
+private theorem relations_liftable {G : Type*} [Group G] {f : B → G} (hf : IsLiftable M f)
+  (r : FreeGroup B) (hr : r ∈ M.relationsSet) : (FreeGroup.lift f) r = 1 := by
+  rcases hr with ⟨⟨i, i'⟩, rfl⟩
+  rw [uncurry, relation, map_pow, _root_.map_mul, FreeGroup.lift.of, FreeGroup.lift.of]
+  exact hf i i'
+
+private def groupLift {G : Type*} [Group G] {f : B → G} (hf : IsLiftable M f) : W →* G :=
+  (PresentedGroup.toGroup (relations_liftable hf)).comp cs.mulEquiv.toMonoidHom
+
+private def restrictUnit {G : Type*} [Monoid G] {f : B → G} (hf : IsLiftable M f) (i : B) :
+    Gˣ where
+  val := f i
+  inv := f i
+  val_inv := pow_one (f i * f i) ▸ M.diagonal i ▸ hf i i
+  inv_val := pow_one (f i * f i) ▸ M.diagonal i ▸ hf i i
+
+/-- Extend the function `f : B → G` to a monoid homomorphism `f' : W → G` satisfying
+`f' (cs.simple i) = f i` for all `i`. -/
+def lift {G : Type*} [Monoid G] {f : B → G} (hf : IsLiftable M f) : W →* G :=
+  MonoidHom.comp (Units.coeHom G) (cs.groupLift
+    (show ∀ i i', ((restrictUnit hf) i * (restrictUnit hf) i') ^ M i i' = 1 from
+      fun i i' ↦ Units.ext (hf i i')))
+
+private theorem toMonoidHom_apply_symm_apply (a : PresentedGroup (M.relationsSet)):
+    (MulEquiv.toMonoidHom cs.mulEquiv : W →* PresentedGroup (M.relationsSet))
+    ((MulEquiv.symm cs.mulEquiv) a) = a := calc
+  _ = cs.mulEquiv ((MulEquiv.symm cs.mulEquiv) a) := by rfl
+  _ = _                                           := by rw [MulEquiv.apply_symm_apply]
+
+theorem lift_apply_simple {G : Type*} [Monoid G] {f : B → G} (hf : IsLiftable M f) (i : B) :
+    cs.lift hf (s i) = f i := by
+  dsimp only [simple, lift, groupLift, MonoidHom.comp_apply]
+  rw [← MonoidHom.toFun_eq_coe, toMonoidHom_apply_symm_apply]
+  simp
+  rfl
```

Comment:
There's a stronger statement available here, right:
```suggestion
private theorem toMonoidHom_apply_symm_apply (a : PresentedGroup (M.relationsSet)) :
    (MulEquiv.toMonoidHom cs.mulEquiv : W →* PresentedGroup (M.relationsSet))
    ((MulEquiv.symm cs.mulEquiv) a) = a := calc
  _ = cs.mulEquiv ((MulEquiv.symm cs.mulEquiv) a) := by rfl
  _ = _                                           := by rw [MulEquiv.apply_symm_apply]

/-- Given a Coxeter group `W` together with a Coxeter system with matrix `M`, indexed by `B`, then
for any monoid `G`, `M`-liftable maps `B → G` are equivalent to monoid homs `W →* G`. -/
def liftEquiv {G : Type*} [Monoid G] : {f : B → G | IsLiftable M f} ≃ (W →* G) where
  toFun f := MonoidHom.comp (Units.coeHom G) (cs.groupLift
    (show ∀ i i', ((restrictUnit f.property) i * (restrictUnit f.property) i') ^ M i i' = 1 from
      fun i i' ↦ Units.ext (f.property i i')))
  invFun f := ⟨f ∘ cs.simple, fun i i' ↦ by
    rw [comp_apply, comp_apply, ← map_mul, ← map_pow, simple_mul_simple_pow, map_one]⟩
  left_inv f := by
    ext i
    simp only [MonoidHom.comp_apply, comp_apply, mem_setOf_eq, groupLift, simple]
    rw [← MonoidHom.toFun_eq_coe, toMonoidHom_apply_symm_apply, PresentedGroup.toGroup.of,
      OneHom.toFun_eq_coe, MonoidHom.toOneHom_coe, Units.coeHom_apply]
    rfl
  right_inv f := by
    -- Some work to do here
    sorry
```
Unfortunately I have very little time now so I cannot do more than leave this brief comment.

Reply:
There's no need to have both `lift` and `liftEquiv` though, right? I mean it's just another syntactic stumbling block for a future `rw` if we have both isn't it?

Reply:
How about I make it an isomorphism but still call it `lift`?